### PR TITLE
⬇️ Merge release v5.2 to dev branch with typeorm syntax updates for feature flag

### LINKED ISSUE #1875: Merge release v5.2 into dev branch (FF)
We need to merge release v5.2 branch into dev branch for release v6.0

### PR DESCRIPTION
**Pull request by ppratikcr7**
## Context
This PR is to resolve #1875. Here we want to pull the typeORM v0.3.0 update related changes to dev branch to release v6.0 with feature flags.

## Summary
Here I have made a new branch `release/v6.0` from `dev` branch and pulled in `release/v5.2` branch. Updated typeorm syntax for feature flag related code.

## Key Changes
- Updated typeORM syntax for feature flag related code in `dev` branch.
- In typeORM v0.3.0 `{ primary: true }` option to define a column as primary key is not supported for relational columns like `@ManyToMany, @OneToMany`. We need to define separate `@PrimaryColumn` and use `@JoinColum` with the relational column. This is done in this PR for feature flag related models.
- In the `typeORMLoader.ts` file, we need to add any new entities/models that we create and also add any new migrations class.
- Created a migration file for the model changes made.

### CODE REVIEW

**Comment by bcb37** [comment]
> * In the `typeORMLoader.ts` file, we need to add any new entities/models that we create and also add any new migrations class.

That's really annoying. Why can't Typeorm keep track of and discover these things anymore? I know I'll forget to do this step a lot.

File: backend/packages/Upgrade/src/loaders/typeormLoader.ts [hunk 1]
```diff
@@ -1,23 +1,187 @@
...
+import { contextInMetric1712553037665 } from '../database/migrations/1712553037665-contextInMetric';
+import { featureFlagFilterMode1714680515570 } from '../database/migrations/1714680515570-featureFlagFilterMode';
+import { addingIndex1716191003726 } from '../database/migrations/1716191003726-addingIndex';
+import { monitoredIndexesForExport1718377498760 } from '../database/migrations/1718377498760-monitoredIndexesForExport';
+import { addTokenValidationFailedEnum1720515933833 } from '../database/migrations/1720515933833-addTokenValidationFailedEnum';
+import { featureFlagLists1720810654183 } from '../database/migrations/1720810654183-feature-flag-lists';
+import { removeGroupIdFromIndividualExclusion1721124249413 } from '../database/migrations/1721124249413-removeGroupIdFromIndividualExclusion';
+import { featureFlagFilterModeChangeDefault1722540825048 } from '../database/migrations/1722540825048-featureFlagFilterModeChangeDefault';
+
+const entities = [
```
Comment (bcb37) [inline backend/packages/Upgrade/src/loaders/typeormLoader.ts:82]: Could we not just use `env.app.dirs.entities` instead of having to define this and update it all the time? I tested it locally and it works.

File: backend/packages/Upgrade/src/loaders/typeormLoader.ts [hunk 2]
```diff
@@ -1,23 +1,187 @@
...
+  ScheduledJob,
+  Segment,
+  Setting,
+  StateTimeLog,
+  StratificationFactor,
+  User,
+  UserStratificationFactor,
+];
+
+export const migrations = [
```
Comment (bcb37) [inline backend/packages/Upgrade/src/loaders/typeormLoader.ts:122]: Same here. I was able to use `env.app.dirs.migrations` and did not have to define this variable.

File: backend/packages/Upgrade/test/utils/database.ts [hunk 3]
```diff
@@ -19,21 +22,42 @@ export const createDatabaseConnection = async (): Promise<Connection> => {
     password: env.db.password,
     logging: env.db.logging as LoggerOptions,
     entities: env.app.dirs.entities,
-    migrations: env.app.dirs.migrations,
-  });
-  return connection;
+    migrations: migrations,
```
Comment (bcb37) [inline backend/packages/Upgrade/test/utils/database.ts:25]: I successfully ran tests using `env.app.dirs.migration` instead of importing the variable and it worked fine.

File: backend/packages/Upgrade/test/utils/database.ts [hunk 4]
```diff
@@ -19,21 +22,42 @@ export const createDatabaseConnection = async (): Promise<Connection> => {
...
+    name: CONNECTION_NAME.REPLICA,
+    type: env.db.type as any, // See createConnection options for valid types
+    database: env.db.database,
+    host: env.db.host,
+    port: env.db.port,
+    username: env.db.username,
+    password: env.db.password,
+    logging: env.db.logging as LoggerOptions,
+    entities: env.app.dirs.entities,
+    migrations: migrations,
```
Comment (bcb37) [inline backend/packages/Upgrade/test/utils/database.ts:38]: Here too

File: backend/packages/Upgrade/package.json [hunk 5]
```diff
@@ -19,9 +19,9 @@
         "seed": "npm start db.seed",
         "drop": "npm start db.drop",
         "help": "npm start help",
-        "migration:generate": "./node_modules/.bin/ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm migration:generate -n",
-        "migration:create": "./node_modules/.bin/ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm migration:create -n",
-        "migration:run": "./node_modules/.bin/ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm migration:run",
+        "migration:generate": "./node_modules/.bin/ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm migration:generate -d src/loaders/typeormLoader.ts",
+        "migration:create": "./node_modules/.bin/ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm migration:create -d src/loaders/typeormLoader.ts",
+        "migration:run": "./node_modules/.bin/ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm migration:run -d src/loaders/typeormLoader.ts",
         "migration:revert": "./node_modules/.bin/ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm migration:revert"
```
Comment (bcb37) [inline backend/packages/Upgrade/package.json:25]: We should add the -d flag to the `revert` script here too

**Review by bcb37 — DISMISSED** [review]
A few things I noticed, which were probably in the original pr.
Also, should the migration in the top level `1724745446870-typeormv3-featureFlagUpdates.ts` be moved?

File: backend/packages/Upgrade/src/api/repositories/AnalyticsRepository.ts [hunk 6]
```diff
@@ -107,12 +108,10 @@ export interface ExperimentCSVData {
   details: ConditionDecisionPointData[];
 }
 
-@EntityRepository()
-export class AnalyticsRepository {
-  constructor(private manager: EntityManager) {}
-
+@EntityRepository(AnalyticsRepository)
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/repositories/AnalyticsRepository.ts:111]: This is inocrrect. Entity should be passes here and becuase this repository doesn't has any Entity nothing should be passed over here

File: backend/packages/Upgrade/1724745446870-typeormv3-featureFlagUpdates.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import { MigrationInterface, QueryRunner } from "typeorm";
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/1724745446870-typeormv3-featureFlagUpdates.ts:1]: Is this file needed?

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]

**Comment by ppratikcr7** [comment]
> > * In the `typeORMLoader.ts` file, we need to add any new entities/models that we create and also add any new migrations class.
> 
> That's really annoying. Why can't Typeorm keep track of and discover these things anymore? I know I'll forget to do this step a lot.

Right, I was also feeling the same as it was done in the old PR for typeorm changes and I was going to look for a solution. Yes, indeed `env.app.dirs.entities` and `env.app.dirs.migration` works fine. Thanks :)

File: backend/packages/Upgrade/1724745446870-typeormv3-featureFlagUpdates.ts (hunk 7 above)
Comment (ppratikcr7) [reply backend/packages/Upgrade/1724745446870-typeormv3-featureFlagUpdates.ts:1]: Yes, somehow the file was created in the Upgrade directory. Will update this.

File: backend/packages/Upgrade/src/api/repositories/AnalyticsRepository.ts (hunk 6 above)
Comment (ppratikcr7) [reply backend/packages/Upgrade/src/api/repositories/AnalyticsRepository.ts:111]: Okay, this change was made in older PR for release 5.2 typeorm changes. I will fix this. Thanks.

File: backend/packages/Upgrade/test/utils/database.ts (hunk 3 above)
Comment (VivekFitkariwala) [reply backend/packages/Upgrade/test/utils/database.ts:25]: Lets go with @bcb37 suggestion till they force us to upgrade :)

**Comment by ppratikcr7** [comment]
@bcb37 @VivekFitkariwala Made changes as requested to remove excess changes to the datasource typeLoader file related to annoying updates to the entities and migration on each new addition. We can update once its forced in typeorm v0.4.

Also, checked the migration scripts and resolved them to generate migration files in proper directory `src/database/migrations`. The change required to run the migration script is as below for the `generate` and `create` cli commands (we need to pass the migration file name as an argument `name`):

`name=migration-name npm run migration:generate
name=migration-name npm run migration:create`

The other two migration scripts for `run` and `revert` should remain the same:

`npm run migration:run
npm run migration:revert`

**Comment by ppratikcr7** [comment]
Once, this is approved, will add these changes in the slack and gitbook documentation.

**Review by VivekFitkariwala — APPROVED** [review]

**Review by VivekFitkariwala — APPROVED** [review]
